torch/nn/parallel/distributed.py

```diff
@@ -509,6 +494,7 @@ def __init__(
     ):
 
         super(DistributedDataParallel, self).__init__()
+        _Joinable.__init__(self)
```

Comment:
Do we need this? Looks like `super().__init__()` should have already covered all parent classes?

```python
class First(object):
    def __init__(self):
        super(First, self).__init__()
        print("first")

class Second(object):
    def __init__(self):
        super(Second, self).__init__()
        print("second")

class Third(First, Second):
    def __init__(self):
        super(Third, self).__init__()
        print("third")

x = Third()
```

outputs

```
second
first
third
```

Reply:
`Module` is not designed for cooperative inheritance since there was probably never a use case before. In particular, `Module.__init__()` does not call `super(Module, self).__init__()`, so `super(DistributedDataParallel, self).__init__()` will call `Module.__init__()` but not continue to call `_Joinable.__init__()`. 

Option 1: Leave it as is.

Option 2: Add `super(Module, self).__init__()` to the end of `Module.__init__()`:
https://github.com/andwgu/pytorch/blob/9a7b6044ac232c9b38ef437fd007a558115bbbd8/torch/nn/modules/module.py#L250-L266

Option 3: Switch the order of inheritance:
```
class DistributedDataParallel(_Joinable, Module):
```
Then, as long as we include `super(_Joinable, self).__init__()` in `_Joinable.__init__()`, then `DistributedDataParallel.__init__()` will call `Module.__init__()`.

Reply:
Option 1 SGTM. Let's also create an issue to track this. I am not sure why `nn.Module` does not call `super().__init__()`.

Reply:
Similar issue with `torch.optim.Optimizer`. I have filed an issue to track this: https://github.com/pytorch/pytorch/issues/61662